### PR TITLE
🔍 💥 Add UX for client-side search

### PR DESCRIPTION
**Pull request by agoose77**
This PR is a rough stab at a frontend component for client-side search. There are already some sharp edges to rework:

- [x] Remove reference to `@myst-theme/search` in `@myst-theme/common` once `mystmd` exports `myst-spec-ext` with the latest tyeps
- [x] Handle pagination or scrolling
- [x] Improve highlight renderer to e.g. reduce ellipsis :rofl: 
- [x] Add "context" e.g. current doc/section and highlight for each entry
- [ ] Think about whether this can/should be generalised for a server-side search

https://github.com/user-attachments/assets/02188866-daf8-4efc-93c7-d1737ff11ed2

> [!NOTE]
> To test this PR:
> 1. Build https://github.com/jupyter-book/mystmd/pull/1530 and run `myst start --headless`
> 2. Run `theme:book` as usual

### CODE REVIEW

**Comment by agoose77** [comment]
@stevejpurves @rowanc1 input always welcome, but I'll keep chipping away at this.

**Comment by agoose77** [comment]
Phew, made some good progress! My goal here is to build out "an approach", and then we can iterate on the design.

**Review by rowanc1 — COMMENTED** [review]
In progress comments!

File: package.json [hunk 1]
```diff
@@ -51,5 +51,11 @@
...
     "node": ">=14.0.0"
   },
-  "packageManager": "npm@8.10.0"
+  "packageManager": "npm@8.10.0",
+  "dependencies": {
+    "@myst-theme/search": "^0.0.0",
+    "@radix-ui/react-visually-hidden": "^1.1.0",
+    "minisearch": "^7.1.0",
+    "string.prototype.matchall": "^4.0.11"
+  }
```
[the remark this answers is not in the document]
Comment (agoose77) [reply package.json:60]: I have realised that the npm command doesn't do what it says it does 🙃

[the remark this answers is not in the document]
Comment (agoose77) [reply package.json:60]: I have realised that the npm command doesn't do what it says it does 🙃

**Review by choldgraf — COMMENTED** [review]
From a user's perspective, based on the two videos above, I think that this is ready to ship and consider it an alpha-level feature. I like the UI responsiveness, and the results modal is a great start - +1 to mimicking the UI of Algolia since they've done a lot of UI research already.

I couldn't test this locally (tried to get this theme building for 15 minutes but couldn't figure it out) but my feeling is that we should try to get this into the hands of users for feedback sooner than later, it already feels ready to me from a functionality perspective.

Is there any major technical blocker that we need to work out first? If not, my vote is to ship this and iterate when we get feedback. I'm happy to start trying it out a bit more.

File: themes/book/app/utils/loaders.server.ts [hunk 2]
```diff
@@ -89,8 +90,9 @@ export async function getMystXrefJson(): Promise<Record<string, any> | null> {
   return xrefs;
 }
 
-export async function getMystSearchJson(): Promise<Record<string, any> | null> {
+export async function getMystSearchJson(): Promise<MystSearchIndex | null> {
   const url = updateLink('/myst.search.json');
+	console.log("REQUESTING SEARCH", url)
```
Comment (agoose77) [inline themes/book/app/utils/loaders.server.ts:95]: Remove this atrocity.
```suggestion
```

**Pull request merged**